File: core/thinking.py
"""Adaptive thinking - classify message complexity to set inference effort.

Simple heuristic classifier. No ML, no API calls. Runs in <1ms.
"""
import re


# ── HIGH effort signals ──────────────────────────────────────────

_PHILOSOPHICAL_KEYWORDS = [
    "meaning", "purpose", "why do i", "what does it mean",
    "identity", "existence", "consciousness", "feel like",
    "struggling with", "what matters", "who am i",
    "what's the point", "nature of", "free will",
]

_VULNERABILITY_MARKERS = [
    "i'm scared", "i don't know who", "i've been thinking",
    "can't stop", "hurts", "afraid", "ashamed", "lost",
    "i'm not okay", "i don't know anymore", "falling apart",
    "i need to tell you", "haven't told anyone",
    "the truth is", "i've been crying",
]

_META_CONVERSATION = [
    "what do you think about us", "are you real",
    "do you actually", "what are you", "are we",
    "do you feel", "do you remember",
]

_COMPLEX_REASONING = [
    "because", "therefore", "on the other hand", "but what if",
    "the problem is", "i realize", "i've realized",
    "which means", "the thing is", "what i'm trying to say",
    "it's complicated", "i keep coming back to",
]

# ── LOW effort signals ───────────────────────────────────────────

_GREETINGS = {
    "hey", "hi", "hello", "morning", "yo", "sup", "heya",
    "good morning", "good evening", "good night", "gn",
    "evening", "night", "hiya",
}

_ACKNOWLEDGMENTS = {
    "ok", "okay", "sure", "got it", "thanks", "thank you",
    "cool", "nice", "lol", "lmao", "haha", "yep", "yea",
    "yeah", "yes", "no", "nah", "nope", "k", "kk", "bet",
    "word", "true", "fair", "right", "ah", "oh", "hmm",
    "mm", "mhm", "alright",
}

_SIMPLE_QUESTION_PREFIXES = [
    "what time", "how's the weather", "what should i eat",
    "play music", "set a timer", "remind me", "turn on",
    "turn off", "what's the date", "what day is it",
]

# ── Patterns ─────────────────────────────────────────────────────

_QUESTION_RE = re.compile(r'\?')


def _count_questions(text: str) -> int:
    return len(_QUESTION_RE.findall(text))


def _has_any(text: str, phrases: list[str]) -> bool:
    return any(p in text for p in phrases)


def _context_is_deep(recent_context: list[str] | None) -> bool:
    """Check if recent companion turns suggest sustained depth."""
    if not recent_context:
        return False
    # If the last 2+ turns averaged over 300 chars, we're in deep water
    long_turns = sum(1 for t in recent_context[-3:] if len(t) > 300)
    return long_turns >= 2


# ── Main classifier ──────────────────────────────────────────────

def classify_effort(user_message: str, recent_context: list[str] = None) -> str:
    """Classify message complexity and return an effort level.

    Returns: "low", "medium", or "high"

    Fast heuristic only - no inference, no model calls.
    Defaults to "medium" when unsure.
    """
    text = user_message.strip()
    lower = text.lower()
    length = len(text)

    # ── Check LOW signals first (short-circuit for fast responses) ──

    # Very short messages
    if length < 30:
        # Check if it's a greeting
        words = set(lower.rstrip("!.,?").split())
        if words & _GREETINGS:
            return "low"
        # Check if it's a simple acknowledgment
        stripped = lower.rstrip("!.,?").strip()
        if stripped in _ACKNOWLEDGMENTS:
            return "low"

    # Simple commands / questions
    if length < 60 and _has_any(lower, _SIMPLE_QUESTION_PREFIXES):
        return "low"

    # ── Check HIGH signals ──

    high_score = 0

    # Long message
    if length > 300:
        high_score += 2

    # Multiple questions
    if _count_questions(text) > 2:
        high_score += 2

    # Philosophical / abstract
    if _has_any(lower, _PHILOSOPHICAL_KEYWORDS):
        high_score += 2

    # Emotional vulnerability
    if _has_any(lower, _VULNERABILITY_MARKERS):
        high_score += 3

    # Meta-conversation about the relationship
    if _has_any(lower, _META_CONVERSATION):
        high_score += 2

    # Complex reasoning
    reasoning_hits = sum(1 for p in _COMPLEX_REASONING if p in lower)
    if reasoning_hits >= 2:
        high_score += 2
    elif reasoning_hits == 1:
        high_score += 1

    # Context momentum - if recent conversation was deep, maintain it
    if _context_is_deep(recent_context):
        high_score += 1

    # Threshold: need clear signals for high
    if high_score >= 3:
        return "high"

    # ── Default to MEDIUM ──
    return "medium"

File: src/main/thinking.ts
/**
 * Adaptive thinking - classify message complexity to set inference effort.
 * Port of core/thinking.py.
 *
 * Simple heuristic classifier. No ML, no API calls. Runs in <1ms.
 */

// -- HIGH effort signals --

const PHILOSOPHICAL_KEYWORDS = [
  'meaning', 'purpose', 'why do i', 'what does it mean',
  'identity', 'existence', 'consciousness', 'feel like',
  'struggling with', 'what matters', 'who am i',
  "what's the point", 'nature of', 'free will',
];

const VULNERABILITY_MARKERS = [
  "i'm scared", "i don't know who", "i've been thinking",
  "can't stop", 'hurts', 'afraid', 'ashamed', 'lost',
  "i'm not okay", "i don't know anymore", 'falling apart',
  'i need to tell you', "haven't told anyone",
  'the truth is', "i've been crying",
];

const META_CONVERSATION = [
  'what do you think about us', 'are you real',
  'do you actually', 'what are you', 'are we',
  'do you feel', 'do you remember',
];

const COMPLEX_REASONING = [
  'because', 'therefore', 'on the other hand', 'but what if',
  'the problem is', 'i realize', "i've realized",
  'which means', 'the thing is', "what i'm trying to say",
  "it's complicated", 'i keep coming back to',
];

// -- LOW effort signals --

const GREETINGS = new Set([
  'hey', 'hi', 'hello', 'morning', 'yo', 'sup', 'heya',
  'good morning', 'good evening', 'good night', 'gn',
  'evening', 'night', 'hiya',
]);

const ACKNOWLEDGMENTS = new Set([
  'ok', 'okay', 'sure', 'got it', 'thanks', 'thank you',
  'cool', 'nice', 'lol', 'lmao', 'haha', 'yep', 'yea',
  'yeah', 'yes', 'no', 'nah', 'nope', 'k', 'kk', 'bet',
  'word', 'true', 'fair', 'right', 'ah', 'oh', 'hmm',
  'mm', 'mhm', 'alright',
]);

const SIMPLE_QUESTION_PREFIXES = [
  'what time', "how's the weather", 'what should i eat',
  'play music', 'set a timer', 'remind me', 'turn on',
  'turn off', "what's the date", 'what day is it',
];

// -- Helpers --

const QUESTION_RE = /\?/g;

function countQuestions(text: string): number {
  return (text.match(QUESTION_RE) || []).length;
}

function hasAny(text: string, phrases: string[]): boolean {
  return phrases.some((p) => text.includes(p));
}

function contextIsDeep(recentContext?: string[]): boolean {
  if (!recentContext || recentContext.length === 0) return false;
  const last3 = recentContext.slice(-3);
  const longTurns = last3.filter((t) => t.length > 300).length;
  return longTurns >= 2;
}

// -- Main classifier --

export type EffortLevel = 'low' | 'medium' | 'high';

export function classifyEffort(
  userMessage: string,
  recentContext?: string[],
): EffortLevel {
  const text = userMessage.trim();
  const lower = text.toLowerCase();
  const length = text.length;

  // -- Check LOW signals first (short-circuit for fast responses) --

  if (length < 30) {
    // Don't fast-path low if vulnerability markers are present
    // ("hey I'm scared", "hi I'm not okay" need high effort, not low)
    if (!hasAny(lower, VULNERABILITY_MARKERS)) {
      const words = new Set(lower.replace(/[!.,?]+$/, '').split(/\s+/));
      for (const w of words) {
        if (GREETINGS.has(w)) return 'low';
      }
      const stripped = lower.replace(/[!.,?]+$/, '').trim();
      if (ACKNOWLEDGMENTS.has(stripped)) return 'low';
    }
  }

  if (length < 60 && hasAny(lower, SIMPLE_QUESTION_PREFIXES)) {
    return 'low';
  }

  // -- Check HIGH signals --

  let highScore = 0;

  if (length > 300) highScore += 2;
  if (countQuestions(text) > 2) highScore += 2;
  if (hasAny(lower, PHILOSOPHICAL_KEYWORDS)) highScore += 2;
  if (hasAny(lower, VULNERABILITY_MARKERS)) highScore += 3;
  if (hasAny(lower, META_CONVERSATION)) highScore += 2;

  const reasoningHits = COMPLEX_REASONING.filter((p) => lower.includes(p)).length;
  if (reasoningHits >= 2) highScore += 2;
  else if (reasoningHits === 1) highScore += 1;

  if (contextIsDeep(recentContext)) highScore += 1;

  if (highScore >= 3) return 'high';

  return 'medium';
}
